jb_toc/jb_toc/titles.py

```diff
@@ -0,0 +1,81 @@
+from __future__ import annotations
+import os, re
+from typing import Dict, Any, List
+from jupyter_server.utils import ensure_async
+from jupyter_server.services.contents.manager import ContentsManager
+
+
+
+def _title_from_markdown(md_text: str) -> str | None:
+    """
+    Searches Markdown for the first heading of any level up to 6.
+    Supports ATX and SETEXT heading styles
+
+    md_text: Markdown text to search for a heading
+    returns: The first encountered heading as a title
+    """
+    atx = re.compile(r'^\s{0,3}#{1,6}\s+(.+?)\s*#*\s*$', re.MULTILINE)
+    setext = re.compile(r'^(.+)\n[=-]{3,}\s*$', re.MULTILINE)
+    result = atx.search(md_text)
+
+    if result:
+        return result.group(1).strip()
+    result = setext.search(md_text)
+
+    if result:
+        return result.group(1).strip()
+    return None
+
+def _title_from_notebook(notebook: Dict[str, Any]) -> str | None:
+    """
+    Searches a notebook's markdown cells for the first heading of any level
+    up to 6 and returns it as a title
+
+    notebook: json object containing the notebook
+    returns: The first encountered heading as a title
+    """
+    title = (notebook.get("metadata") or {}).get("title")
+
+    if title:
+        return str(title)
+    for cell in notebook.get("cells", []):
+        if cell.get("cell_type") == "markdown":
```

Comment:
nit: you could lose a level of indentation here with

```suggestion
        if not cell.get("cell_type") == "markdown":
            continue
```